### ISSUE TITLE
Add interrupt annotations

### ISSUE THREAD

**Issue by mysterymath** [llvm-mos/llvm-mos#66]
Clang annotations should be added to support interrupt handling on MOS.

For interrupts to work with the static stack optimization, two function attributes are needed: `__attribute__((interrupt))` and `__attribute__((interrupt_norecurse))`. It will be undefined behavior for an interrupt to call a C function that does not bear one of these two annotations. The only real difference between the two is that if a an interrupt enters a `interrupt_norecurse` function, it is undefined behavior for another interrupt to enter that function while the first invocation is active. `interrupt` has no such requirements. These correspond to roughly the "fast" and "slow" portions of interrupt handlers.

These attributes will cause all necessary ISR save and restore code to be inserted into the annotated function, and the function will return with `RTI` instead of `RTS`. This can be skipped by adding another attribute, `__attribute__((no_isr))`. This indicates that the function is the C entry point for an assembly-language ISR that manages saving and restoring registers. Accordingly, the regular C calling convention will be used, but the other semantics of `interrupt` and `interrupt_norecurse` still apply. It shall be an error to apply this attribute to a function that is not marked as `interrupt` or `interrupt_norecurse`.

Use of any of these attributes anywhere in a program will globally modify the calling convention and use of zero page registers. The number of available zero page pointers will be capped at 5. This ensures that the ISR can execute in a timely fashion, as it must then save and restore 10 - 4 = 6 bytes of zero page pointers (4 bytes are callee-saved and don't need to be saved in ISRs), as well as 5 compiler temporary locations, for a total of 11 LDA PHA pairs, along with the usual pushing of A, X, and Y.

The MOSRecurse pass will need to be modified to walk the call graphs rooted at `interrupt` and `interrupt_norecurse` functions. The following are marked as possibly recursive:

- Any function possibly callable by an `interrupt` function.
- Any function callable by main and an `interrupt_norecurse` function, or by two different `interrupt_norecurse` functions.

This is far from ideal, but it's a starting point. Judicious use of `interrupt_norecurse` will allow carefully designed interrupt handlers to use the static stack. This, along with 5 zero page pointers, should be enough to allow writing working, but highly suboptimal, code.

Later optimization work will improve usage of zero page registers in the presence of C interrupt handlers.

**Comment by johnwbyrd** [comment]
I keep thinking about this problem from the following perspective: what would I do, if I needed to write an interrupt in C, and the 6502 C compiler didn't support interrupt concepts?

In that case, I would try to convince the compiler to emit code for a similar target with a similar linker script, but with a different base address and range for the imaginary registers, so that the interrupt code would not collide with the main code.

I'd still be personally responsible for choosing those imaginary register ranges, telling the compiler about them in each invocation, and working around -flto differences from link session to link session.

In short, I think it is conceptually possible to implement interrupts by setting up two banks of imaginary registers, and letting the compiler register allocation on both banks independently, based on the current __attribute___ setting.

I am not recommending this idea -- there are plenty of flaws in it.  It pushes responsibility for managing these banks onto the linker author.  All bets are off if you call malloc() -- although, most interrupt programmers are already ok with this restriction.  I think you should try the method you are describing.  No idea how to dance around -flto differences -- there would presumably need to be multiple link steps, but I don't even know if that would work.  

tl:dr; go with plan A.  If plan A does not work, all is not necessarily lost; maybe we can get the linker involved and piece together a worse plan B.

**Comment by mysterymath** [comment]
The big issue with this approach is that interrupt handlers usually need to touch some of the same data structures as main or other interrupts; otherwise the interrupts really can't be said to do anything. To have fully separate linking like this, all such data structures would need to be given fixed memory locations outside of the purview of the linker; this is quite a burden!

More concerning is that you'd also need to maintain N complete sets of libcalls and compiler temporaries, one for each interrupt domain. This could easy blow out the space available to run a program.

I've come to realize that nearly all the difficulty in interrupts isn't in how to keep things separate, it's how to keep only certain things separate. We really do want things like compiler temporaries and libcalls to be usable from multiple contexts, *and* to use the zero page well. It's a hard problem, but I'm becoming increasingly confident that we can't really dodge it.
